### PR TITLE
[2/2] Intel GPU Runtime Upstreaming for Generator

### PR DESCRIPTION
**Pull request by guangyey**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120602
* __->__ #118613

# Motivation
According to [[1/2] Intel GPU Runtime Upstreaming for Generator](https://github.com/pytorch/pytorch/pull/118528), as mentioned in [[RFC] Intel GPU Runtime Upstreaming](https://github.com/pytorch/pytorch/issues/114842), the second PR covers the changes under `python frontend`.

# Design
Currently, it primarily offers geneartor-related APIs, including

- `torch.xpu.default_generators`
- `torch.xpu.get_rng_state`
- `torch.xpu.get_rng_state_all`
- `torch.xpu.initial_seed`
- `torch.xpu.manual_seed`
- `torch.xpu.manual_seed_all`
- `torch.xpu.seed`
- `torch.xpu.seed_all`
- `torch.xpu.set_rng_state`
- `torch.xpu.set_rng_state_all`

# Additional Context
The differences with CUDA:
The generator-related frontend python APIs are 1:1 mapping with CUDA.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by gujinghui — APPROVED** [review]

**Review by EikanWang — APPROVED** [review]

File: aten/src/ATen/detail/XPUHooksInterface.h [hunk 1]
```diff
@@ -55,6 +51,10 @@ struct TORCH_API XPUHooksInterface {
     return 0;
   }
 
+  virtual DeviceIndex current_device() const {
+    return -1;
+  }
```
Comment (jgong5) [inline aten/src/ATen/detail/XPUHooksInterface.h:56]: This seems related to device, not generator.

**Review by jgong5 — APPROVED** [review]

File: aten/src/ATen/detail/XPUHooksInterface.h (hunk 1 above)
Comment (guangyey) [reply aten/src/ATen/detail/XPUHooksInterface.h:56]: This is used in `manual_seed` in `xpu.cpp` file.

File: test/test_utils.py [hunk 2]
```diff
@@ -822,17 +815,6 @@ def test_external_module_register(self):
         with self.assertRaisesRegex(RuntimeError, "Expected one of cpu"):
             torch._register_device_module('dummmy', DummyXPUModule)
 
-        with self.assertRaises(AttributeError):
```
Comment (guangyey) [inline test/test_utils.py:825]: Now `torch.xpu` is a built-in module. We remove these tests as they are covered by `test_external_module_and_backend_register`. And remove `tearDown` because it is unnecessary anymore.

File: aten/src/ATen/detail/XPUHooksInterface.h [hunk 3]
```diff
@@ -55,6 +51,10 @@ struct TORCH_API XPUHooksInterface {
     return 0;
   }
 
+  virtual DeviceIndex current_device() const {
+    return -1;
```
Comment (albanD) [inline aten/src/ATen/detail/XPUHooksInterface.h:55]: I know the value here is copy pasted from the cuda code but why is the current device -1 when XPU is not available? Why wouldn't that be an error like every other default impl here?

File: aten/src/ATen/detail/XPUHooksInterface.h [hunk 4]
```diff
@@ -43,10 +43,6 @@ struct TORCH_API XPUHooksInterface {
     TORCH_CHECK(false, "Cannot get XPU global device index without ATen_xpu library.");
   }
 
-  virtual Generator getXPUGenerator(C10_UNUSED DeviceIndex device_index = -1) const {
```
Comment (albanD) [inline aten/src/ATen/detail/XPUHooksInterface.h:46]: Why is this removed?

File: torch/csrc/Generator.cpp [hunk 5]
```diff
@@ -72,9 +76,12 @@ static PyObject* THPGenerator_pynew(
     self->cdata = make_generator<MPSGeneratorImpl>();
   }
 #endif
+#ifdef USE_XPU
   else if (device.type() == at::kXPU) {
-    self->cdata = at::detail::getXPUHooks().getXPUGenerator(device.index());
-  } else if (device.type() == at::kIPU) {
+    self->cdata = make_generator<XPUGeneratorImpl>(device.index());
```
Comment (albanD) [inline torch/csrc/Generator.cpp:81]: The old version via hooks and less ifdef sounds better?

File: torch/csrc/xpu/Module.cpp [hunk 6]
```diff
@@ -272,6 +273,20 @@ static PyObject* THXPModule_initExtension(PyObject* self, PyObject* noargs) {
...
     throw python_error();
 
+  auto set_module_attr = [&](const char* name, PyObject* v) {
+    if (PyObject_SetAttrString(m, name, v) < 0) {
+      throw python_error();
+    }
+  };
+
+  auto num_gpus = c10::xpu::device_count();
+  auto default_xpu_generators = PyTuple_New(static_cast<Py_ssize_t>(num_gpus));
```
Comment (albanD) [inline torch/csrc/xpu/Module.cpp:283]: I would discourage the use of "auto" for CPython API.
I think you're leaking this object here.

File: torch/xpu/__init__.py [hunk 7]
```diff
@@ -16,11 +17,40 @@
...
+_tls = threading.local()
 _initialization_lock = threading.Lock()
+_queued_calls: List[
+    Tuple[Callable[[], None], List[str]]
+] = []  # don't invoke these until initialization occurs
 _is_in_bad_fork = getattr(torch._C, "_xpu_isInBadFork", lambda: False)
 _device_t = Union[_device, str, int, None]
 
 
+class _LazySeedTracker:
```
Comment (albanD) [inline torch/xpu/__init__.py:29]: Can we factor that out to not duplicate the impl?

File: aten/src/ATen/detail/XPUHooksInterface.h (hunk 3 above)
Comment (guangyey) [reply aten/src/ATen/detail/XPUHooksInterface.h]: Done. Raising an error looks better.

File: aten/src/ATen/detail/XPUHooksInterface.h (hunk 4 above)
Comment (guangyey) [reply aten/src/ATen/detail/XPUHooksInterface.h]: Change back. Using hooks instead of ifdef.

File: torch/csrc/Generator.cpp (hunk 5 above)
Comment (guangyey) [reply torch/csrc/Generator.cpp]: Using hooks instead of ifdef.

File: torch/xpu/__init__.py (hunk 7 above)
Comment (guangyey) [reply torch/xpu/__init__.py:29]: Is it reasonable to refactor these code in an additional PR, see [refactor code to share across different devices](https://github.com/pytorch/pytorch/pull/120602). When upstreaming, we would like not to affect the CUDA code.

**Review by albanD — COMMENTED** [review]
The code change sounds fine but the CI failures are real?

**Comment by guangyey** [comment]
@albanD All CI pass after I rebase to the latest main branch.

**Review by albanD — APPROVED** [review]
Sounds good!
